packages/SwingSet/src/vats/comms/state.js

```diff
@@ -244,12 +244,20 @@ export function makeStateKit(state) {
     assert(!p.resolved);
     insistCapData(resolution.data);
     p.resolved = true;
+    p.kernelAwaitingResolve = true;
```

Comment:
I think this does the right thing, so let's land it, but the model it presents could be clearer, so let's plan to improve this logic later. Plus, I'd really like to get rid of my ugly `Promise.resolve().then` hack.

The right way to think about this is as if there were a full c-list between comms and the kernel, as if the comms-numberspace promise ID was not always the same as the kernel-facing `vpid`. We actually use a degenerate c-list, in which those identifiers really are always the same, except that the `vpid` isn't always in the c-list. If one remote sends a message (which contains some new promise as an argument) to a different remote that passes through our comms vat (via an unshortened third-party proxy), our comms vat will allocate a promise ID for it, but won't have sent it to the kernel in any syscall.

Now two things might happen, in either order:

* the decider of that promise sends us a resolution for it
* someone who knows that promise sends a message to our kernel that mentions it

The first causes us to set the `p.resolved` flag. The second should cause us to set a flag that says "the kernel knows about this, and needs to be notified if/when it gets resolved". (I'm 80% confident that the latter state is exactly equal to the `p.kernelIsSubscribed` flag, so maybe we can re-use it).

If the kernel is subscribed at the time that the promise gets resolved, we need to do a `syscall.resolve`, and then clear the flag. If the promise ID is translated to kernel-facing space for the first time, we can use the allocated comms-space ID as-is. If the promise ID is translated to kernel-facing space for a second time (because we did a syscall.resolve earlier, and retired the ID), then we need to make a new ID for it.

To get rid of the `Promise.resolve().then`, I think we need a second data pathway from `provideKernelForLocal` out to `resolveToKernel` or `sendToKernel`, like we effectively do in liveslots. I like the idea of the "super-duper serializer" which returns both the primary capdata and a batch of resolutions that need to be delivered: we'd put the extra resolutions in this batch, instead of scheduling them for later within the same crank with `Promise.resolve().then`.

Reply:
moved to #2363